parquet-avro/src/main/java/parquet/avro/AvroReadSupport.java

```diff
@@ -59,6 +70,115 @@ public static void setAvroDataSupplier(Configuration configuration,
+    Map<String, Set<String>> unmergedKeyValueMetadata = context.getKeyValueMetadata();
+    Map<String, String> mergedKeyValueMetadata = new HashMap<String, String>();
+    Iterator<Map.Entry<String, Set<String>>> entryIterator = unmergedKeyValueMetadata.entrySet().iterator();
+
+    while (entryIterator.hasNext()) {
+      Map.Entry<String, Set<String>> entry = entryIterator.next();
+      String key = entry.getKey();
+      if (key.equals(AVRO_SCHEMA_METADATA_KEY)) {
+        // check compatibility of avro schemas in metadata
+        Set<String> writerSchemaStrings = entry.getValue();
+        Set<Schema> writerSchemas = new HashSet<Schema>();
+        /* It is necessary to instantiate a Schema.Parser for each schema
+         * because we are parsing for the same type with different, "evolved" definitions!
+         */
+        for (String s : writerSchemaStrings) {
+          Schema.Parser parser = new Schema.Parser();
+          Schema writerSchema = parser.parse(s);
+          writerSchemas.add(writerSchema);
+        }
+        if (!areSchemasCompatible(readerSchema, writerSchemas))
+          throw new RuntimeException("could not merge metadata: key " + AVRO_SCHEMA_METADATA_KEY + " contains incompatible schemas");
+        // if all writer schemas are compatible with the reader, reassign metadata schema to the reader schema
+        mergedKeyValueMetadata.put(AVRO_SCHEMA_METADATA_KEY, readerSchema.toString());
+      } else if (entry.getValue().size() > 1) {
+        throw new RuntimeException("could not merge metadata: key " + key  + " has conflicting values");
+      } else {
+        String value = entry.getValue().iterator().next();
+        mergedKeyValueMetadata.put(key, value);
+      }
+    }
+    return mergedKeyValueMetadata;
+  }
+
+  /**
+   * If an {@link #AVRO_READ_SCHEMA} is provided and the {@link #AVRO_SCHEMA_COMPATIBILITY_CHECK}
+   * flag has been set, all Avro schemas found in the metadata will be checked for compatibility
+   * against the {@code AVRO_READ_SCHEMA}.
+   */
+  @Override
+  public ReadContext init(InitContext context) {
+    Configuration configuration = context.getConfiguration();
+    String readerSchemaString = configuration.get(AVRO_READ_SCHEMA);
+    boolean checkSchemaCompatibility = configuration.getBoolean(AVRO_SCHEMA_COMPATIBILITY_CHECK, false);
+    // attempt to resolve/evolve writer schemas to the reader schema
+    // iff a reader schema is provided and the compatibility check flag is flipped
+    if (readerSchemaString != null && checkSchemaCompatibility) {
+      Schema.Parser parser = new Schema.Parser();
+      Schema readerSchema = parser.parse(readerSchemaString);
+      return init(configuration, mergeKeyValueMetadata(context, readerSchema), context.getFileSchema());
+    // otherwise, do not attempt to perform schema resolution/evolution
+    } else {
+      return init(configuration, context.getMergedKeyValueMetaData(), context.getFileSchema());
```

Comment:
InitContext#getMergedKeyValueMetaData() is deprecated, so it would be better to avoid calling it. How about always calling a private mergeKeyValueMetadata(context) method that only checks compatibility if checkSchemaCompatibility is true, otherwise just falls back to the behaviour in InitContext#getMergedKeyValueMetaData().